doc/developer-guide/plugins/hooks-and-transactions/ssl-hooks.en.rst

```diff
@@ -84,6 +84,29 @@ handshake processing will not proceed until :c:func:`TSSslVConnReenable()` is ca
 It may be useful to delay the TLS handshake processing if other resources must be consulted to select or create
 a certificate.
 
+TS_SSL_VERIFY_CLIENT_HOOK
+-------------------------
+
+This hook is called when a client connects to Traffic Server and presents a 
+client certificate in the case of a mutual TLS handshake.  The callback can
+get the SSL object from the TSVConn argument and use that to access the client
+certificate and make any additional checks.
+
+Processing will continue regardless of whether the hook callback executes
+:c:func:`TSSslVConnReenable()` since the openssl implementation does not allow
+for pausing processing during the certificate verify callback.
+
+TS_SSL_VERIFY_SERVER_HOOK
+-------------------------
+
+This hooks is called when a Traffic Server connects to an origin and the origin
+presents a certificate.  The callback can get the SSL object from the TSVConn
+argument and use that to access the origin certificate and make any additional checks.
+
+Processing will continue regardless of whether the hook callback executes
```

Comment:
Correct.  The underlying openssl structure limits what we can do.  However, in the case of the verify_server hook, we cannot leave the callback not-enabled.  But we can return an ERROR and have the core ATS verify_callback interpret that to fail the certificate verification.

Reply:
>  The underlying openssl structure limits what we can do

Just to make sure we are on the side page. We agree that if the callback function from `SSL_set_verify()` returns 0, the OpenSSL will stop the TLS handshake, right? So what do you mean here?

> The return value of verify_callback controls the strategy of the further verification process. If verify_callback returns 0, the verification process is immediately stopped with "verification failed" state

Reply:
Correct.  The verify_callback() function is registered by Traffic Server via the SSL_CTX_set_verify or SSL_set_verify openssl function.  It will eventually return either 0 or 1.  If it returns 0, the handshake will fail . It it returns 1, the handshake will continue on.

The VERIFY_SERVER_HOOK gives a place for plugins to insert logic and influence the return value of the verify_callback().  If the hook reenables with a TS_EVENT_ERROR, the verify_callback() will return 0.